### PR TITLE
[5pt] PR: Format and Style Repository

### LINKED ISSUE #34: [2pt] PEP 8?
Considering this repo will be open to the public, I think PEP 8 styling would be appropriate. [Black](https://github.com/psf/black) makes this easy. There are some pitfalls when using git blame after the change, but there are [workarounds](https://github.com/psf/black#migrating-your-code-style-without-ruining-git-blame).

This is simple to do, but it would touch all python code in the project. So it has a pretty big impact to the repo.

### LINKED ISSUE #194: [8pt] Standardize code
Formatting and linting code to be consistent and conform with standards (such as PEP8)

### LINKED ISSUE #846: Add line breaks and/or incorporate linting
Incorporating the use of a code linter would greatly benefit the layout and formatting of this repository.

In the absence of a linter, breaking up long lines of code should be tackled for readability's sake. 

## Current code

No line breaks exist in many bash scripts and python files. This proves difficult to read, understand, and modify arguments passed. 

## Expected code

An attempt should be made to adhere to the [PEP-8 Standard](https://peps.python.org/pep-0008/#:~:text=The%20Python%20standard%20library%20is,inside%20parentheses%2C%20brackets%20and%20braces.) of 79 characters (docstrings/comments to 72). 

While there is no definitive style guide for Shell Scripts, the [Google Shell Style Guide](https://google.github.io/styleguide/shellguide.html#s5.2-line-length-and-long-strings) can be used a starting point.

### LINKED ISSUE #961: Python Black, Flake8, & isort
Look into adding black as a formatter for our codebase.

### PR DESCRIPTION
**Pull request by robgpita-noaa**
Format code using `black` formatter, incorporate `isort` package to sort Python import statements,
and adhere all code to PEP8 Style Guide using the `flake8` package. Set line length to 110 characters. Remove deprecated files. 
Set up git pre-commit hooks.

Will Close Issues:
Closes #34 , Closes #194 , Closes #846 & Closes #961

Will Close PRs:
Resolves #984 , Resolves #979 

Not all files were modified, however, to avoid individually listing each file here, the `/*` convention
is used to denote that almost every file in those directories were formatted and linted.

### Changes

- `.gitattributes`: Add newline at EOF.
- `.github/*`: 
- `.gitignore`: Trim extra last line.
- `CONTRIBUTING.md`: Update contributing guidelines.
- `Dockerfile`: Update PYTHONPATH to point to correct `unit_tests` directory.
- `Pipfile`: Add flake8, black, pyproject-flake8, pre-commit, isort packages
- `Pipfile.lock`: Update to correspond with new packages in Pipfile 
- `README.md` : Update link to wiki, trim whitespace.
- `config/*`
- `data/*`
- `docs/*`
- `fim_pipeline.sh` : Clean up usage statement
- `fim_post_processing.sh`: Update usage statement
- `fim_pre_processing.sh`: Update usage statement.
- `fim_process_unit_wb.sh`: Make usage functional, combine usage and comments.
- `src/*`
- `tools/*`
- `unit_tests/`: The directory name where the unit test data must reside was changed from  `fim_unit_test_data_do_not_remove` => `unit_test_data`

### Additions

- `pyproject.toml`: Configuration file
- `.pre-commit-config.yaml`: Initialize git pre-commit hooks

### Removals

- `data/nws/preprocess_ahps_nws.py`
- `src/adjust_headwater_streams.py`
- `src/aggregate_vector_inputs.py`
- `src/utils/reproject_dem.py`
- `tools/code_standardizer/*`: Incorporated "code_standardizer" into base level Dockerfile.
- `tools/compile_comp_stats.py`
- `tools/compile_computational_stats.py`
- `tools/consolidate_metrics.py`
- `tools/copy_test_case_folders.py`
- `tools/cygnss_preprocessing.py`
- `tools/nesdis_preprocessing.py`
- `tools/plots/*`: Duplicate and unused directory.
- `.isort.cfg`: Incorporated into `pyproject.toml`

### Testing

Ran two alpha domains using fim_4_4_0_1 (latest dev branch) and this branch.  Used the newly added `tools/hash_compare.py` script to compare the hashes of all `.csv` & `.tif` files at the branch level in for all HUCs in the alpha domain. There were no differences in hashes. 

### Issuer Checklist (For developer use)

_You may update this checklist before and/or after creating the PR. If you're unsure about any of them, please ask, we're here to help! These items are what we are going to look for before merging your code._

- [x] Informative and human-readable title, using the format: `[_pt] PR: <description>`
- [x] Links are provided if this PR resolves an issue, or depends on another other PR
- [x] If submitting a PR to the `dev` branch (the default branch), you have a descriptive [Feature Branch](https://www.atlassian.com/git/tutorials/comparing-workflows/feature-branch-workflow) name using the format: `dev-<description-of-change>` (e.g. `dev-revise-levee-masking`)
- [x] Changes are limited to a single goal (no scope creep)
- [x] The feature branch you're submitting as a PR is up to date (merged) with the latest `dev` branch
- [x] Changes adhere to [PEP-8 Style Guidelines](https://peps.python.org/pep-0008/)
- [x] Any _change_ in functionality is tested
- [ ] Passes all unit tests locally (inside interactive Docker container, at `/foss_fim/`, run: `pytest unit_tests/`)
 **   breaking changes of `test_usgs_gage_crosswalk` were introduced in fim4_4_0_0 **
- [ ] New functions are documented (with a description, list of inputs, and expected output)
- [x] Placeholder code is flagged / future todos are captured in comments
- [x] [CHANGELOG](/docs/CHANGELOG.md) updated with template version number, e.g. `4.x.x.x` 
- [x] [Reviewers requested](https://help.github.com/articles/requesting-a-pull-request-review/)
- [x] Add yourself as an [assignee](https://docs.github.com/en/issues/tracking-your-work-with-issues/assigning-issues-and-pull-requests-to-other-github-users) in the PR  as well as the FIM Technical Lead

### Merge Checklist (For Technical Lead use only)

- [x] Update [CHANGELOG](/docs/CHANGELOG.md) with latest version number
- [x] Update the [Citation.cff](/CITATION.cff) file to reflect the latest version number in the [CHANGELOG](/docs/CHANGELOG.md)
- [ ] If applicable, update [README](/README.md) with major alterations

### CODE REVIEW

File: src/add_crosswalk.py [hunk 1]
```diff
@@ -1,264 +1,383 @@
...
     input_src_base = input_src_base.rename(columns=lambda x: x.strip(" "))
-    input_src_base = input_src_base.apply(pd.to_numeric,**{'errors' : 'coerce'})
-    input_src_base['TopWidth (m)'] = input_src_base['SurfaceArea (m2)']/input_src_base['LENGTHKM']/1000
-    input_src_base['WettedPerimeter (m)'] = input_src_base['BedArea (m2)']/input_src_base['LENGTHKM']/1000
-    input_src_base['WetArea (m2)'] = input_src_base['Volume (m3)']/input_src_base['LENGTHKM']/1000
-    input_src_base['HydraulicRadius (m)'] = input_src_base['WetArea (m2)']/input_src_base['WettedPerimeter (m)']
+    input_src_base = input_src_base.apply(pd.to_numeric, **{'errors': 'coerce'})
+    input_src_base['TopWidth (m)'] = input_src_base['SurfaceArea (m2)'] / input_src_base['LENGTHKM'] / 1000
+    input_src_base['WettedPerimeter (m)'] = (
+        input_src_base['BedArea (m2)'] / input_src_base['LENGTHKM'] / 10009
```
Comment (robgpita-noaa) [inline src/add_crosswalk.py:294]: @CarsonPruitt-NOAA This extra 9 was causing the incorrect checksums I was seeing earlier today. Seems to be resolved as of the latest commit.

Comment (CarsonPruitt-NOAA) [reply src/add_crosswalk.py]: Oh wow that was a really good catch. That's a huge typo that I'm sure was affecting our synthetic rating curves. It's in our Manning's equation calculation!

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 2]
```diff
@@ -19,16 +19,17 @@ _You may update this checklist before and/or after creating the PR. If you're un
 - [ ] Changes are limited to a single goal (no scope creep)
 - [ ] The feature branch you're submitting as a PR is up to date (merged) with the latest `dev` branch
 - [ ] Changes adhere to [PEP-8 Style Guidelines](https://peps.python.org/pep-0008/)
```
Comment (mluck) [inline .github/PULL_REQUEST_TEMPLATE.md:21]: With the auto-formatters and linters and pre-commit hooks installed by this PR, is this line necessary anymore?

Comment (robgpita-noaa) [reply .github/PULL_REQUEST_TEMPLATE.md:21]: Good catch. No need for that line anymore, I'll remove it in my next commit. Thanks.

**Comment by RobHanna-NOAA** [comment]
Structurally, the linting tools and hash_compare script look good to me. If the pages and outputs are testing out fine, then I am ok  with it being approved.

**Review by mluck — APPROVED** [review]
Hash comparison between v4.4.0.1 and dev-format-and-style passed after the latest commit.

However, I did discover a slightly strange behavior in an earlier commit where the hash comparison failed for HUC 02030105 with jobs submitted with `-jh 1 -jh 6` (vs `-jh 2 -jb 3`) related to rounding of the 10th decimal place in one of the `hydroTable_0.csv` values.

File: src/src_roughness_optimization.py [hunk 3]
```diff
@@ -148,232 +286,431 @@ def update_rating_curve(fim_directory, water_edge_median_df, htable_path, huc, b
...
-                        continue # if above conditions are True than stop traversing downstream and move on to next starting hydroid
+                    # check if the NextDownID is referenced by more than one hydroid
+                    # (>1 means this is a confluence)
+                    check_confluence = (df_input_htable.NextDownID == nextid).sum() > 1
+                    nextorder = df_input_htable.loc[
+                        nextid, 'order_'
+                    ]  # find the streamorder for the next downstream hydroid
+                    # check if the nextdownid streamorder is greater than the current hydroid order and the
+                    # nextdownid is a confluence (more than 1 upstream hydroid draining to it)
+                    if nextorder > order and check_confluence == True:
```
Comment (robgpita-noaa) [inline src/src_roughness_optimization.py:616]: @CarsonPruitt-NOAA @RyanSpies-NOAA @mluck - I just confirmed that this line was indeed the cause of the difference in hashes we experienced earlier. For reference:

`if nextorder > order and check_confluence is True:` was the previous version that altered the behavior.

**Review by CarsonPruitt-NOAA — APPROVED** [review]

**Pull request merged**